kivy/config.py

```diff
@@ -539,6 +541,7 @@ def getdefaultint(self, section, option, defaultvalue):
     def adddefaultsection(self, section):
         '''Add a section if the section is missing.
         '''
+        assert("_" not in section)
```

Comment:
You don't need parens around an `assert` statement!

Overall, I don't think this should be an `assert` statement at all, because you shouldn't be using `assert`s to validate user input data.

The reason is simple - asserts don't go off if you pass the `-O` flag to Python, and in this case, it means you won't check this condition at all.

So if you were going to keep this test, perhaps consider making it throw a `ValueError`.

---

However, can't you do better than that even?  Adding another arbitrary condition that someone needs to remember is not just inelegant, it adds a little extra overhead for each developer.  Yes, it's tiny here, but these things mount up.

Reading this fragment of code makes it seem as if you can add arbitrary sections just by naming them, and you get no warning at all.

This means I can make a mistake like mistyping `KCFG_GRAPHICS_FULLSCREEN` and get no visible error:  for example

    KCFG_GRAPHIC_FULLSCREEN=1 my_program

Isn't there some way to detect the list of valid sections, and conceivably even valid names within that section, perhaps in kivy/config.py?  

Then you could peel off the section name whether or not it had an `_` in it, and as a bonus could also catch all sorts of niggling errors.

So if you did that, you could get rid of the `assert` entirely, _and_ catch more errors, _and_ present a clearer error message when you did.

Just a thought.  I realize I'm proposing work for someone else. :-)  

Although, I have some spare time this week if you needed it knocked out...

Thanks again for doing this!

Reply:
In my mind, assert _ was just here because it's mostly a reminder for us (dev) not for the user. The user don't have to touch the kivy configuration file, as it has its own application configuration. This is why i did the assert compared to raising an exception.

I could check if the section is valid, but i can't check if the name is valid. Because there is some sections like "input" where the key is up to the user. So there is no validation for that. I'll update the implementation.

Reply:
I second some of the concerns of @rec and would emphasis on not using parenthesis on `assert` because it's a keyword rather than a function. Using parenthesis may lead to unexpected behavior in certain conditions